src/caip-types.ts

```diff
@@ -28,83 +30,95 @@ export const CAIP_ASSET_ID_REGEX =
 /**
  * A CAIP-2 chain ID, i.e., a human-readable namespace and reference.
  */
-export const CaipChainIdStruct = pattern(
-  string(),
+export const CaipChainIdStruct = definePattern<`${string}:${string}`>(
+  'CaipChainId',
   CAIP_CHAIN_ID_REGEX,
-) as Struct<CaipChainId, null>;
```

Comment:
In the PR description you say that the struct is being inferred as a string, but it looks like we were using a type assertion before to ensure that this was not the case. So, was it really being inferred as a string? It's not clear to me what effects these changes have, as the only tests being added in this PR are for `definePattern` and not for any of the structs.

Reply:
`pattern(string(), ...)` would result in a type like this `Struct<string, ...>`

Meaning that using `Infer<typeof MyType>` would be inferred as `string`.

With the new `definePattern` helper I've added, we now uses a generic type to inject it into the `Struct` like so:

```ts
export const CaipChainIdStruct = pattern(
  string(),
  CAIP_CHAIN_ID_REGEX,
); // Would be: Struct<string, null> (which is why we were type-casting)

export const CaipChainIdStruct = definePattern<`${string}:${string}`>(
  'CaipChainId',
  CAIP_CHAIN_ID_REGEX,
); // Would be: Struct<`${string}:${string}`, null> (no more type-casting required)
```

In addition to that, we use `define` under the hood, which allows use to have a "named" struct (and better error messages).